fulltext/__init__.py

```diff
@@ -1,339 +1,166 @@
+    pass
 
-    As mentioned, this is a contrived example. You would be better served by the following.
 
-    >>> import fulltext
-    >>> fulltext.add_handler('text/plain', fulltext.read_content)
+def _get_path(backend, path, **kwargs):
+    """
+    Handle a path.
 
-    But, since the default action is to read the content of an unrecognized file type,
-    registering text/plain is redundant.
+    Called by `get()` when provided a path. This function will prefer the
+    backend's `_get_path()` if one is provided Otherwise, it will open the
+    given path then use `_get_file()`.
     """
-    assert callable(handler), 'Handler must be callable.'
-    FUNC_MAP[(mime, encoding)] = handler
+    if callable(getattr(backend, '_get_path', None)):
+        return backend._get_path(path, **kwargs)
 
+    elif callable(getattr(backend, '_get_file', None)):
+        with open(path, 'r') as f:
+            return backend._get_file(f, **kwargs)
 
-def add_type(mime, ext):
-    """
-    Adds a new mime type and associated extension. This just dispatches to the mimetypes
-    module for now. It is here only to allow future expansion (if we don't use mimetypes
-    forever.)
-    """
-    assert ext.startswith('.'), 'Extension should start with a period `.`.'
-    return mimetypes.add_type(mime, ext)
+    else:
+        BackendError('Backend must provide `_get_path()` or `_get_file()`')
 
 
-def add(mime, ext, handler, commands=None, encoding=None):
+def _get_file(backend, f, **kwargs):
     """
-    Shortcut command to add type, handler and optionally commands. Simply calls add_type(),
-    add_handler() and add_commands().
+    Handle a file-like object.
+
+    Called by `get()` when provided a file-like. This function will prefer the
+    backend's `_get_file()` if one is provided. Otherwise, it will write the
+    data to a temporary file and call `_get_path()`.
     """
-    add_type(mime, ext)
-    add_handler(mime, handler, encoding=encoding)
-    if commands:
-        add_commands(mime, commands, encoding=encoding)
+    if callable(getattr(backend, '_get_file', None)):
+        return backend._get_file(f, **kwargs)
 
+    elif callable(getattr(backend, '_get_path', None)):
+        LOGGER.warning("Using disk, backend does not provde `_get_file()`")
 
-def get_type(filename):
-    """
-    Gets the mimetype and encoding using the mimetypes module. Defined as a standalone
-    function for future expansion.
-    """
-    return mimetypes.guess_type(filename)
+        ext = ''
+        if 'ext' in kwargs:
+            ext = '.' + kwargs['ext']
 
+        with tempfile.NamedTemporaryFile(dir=FULLTEXT_TEMP, suffix=ext) as t:
+            shutil.copyfileobj(f, t)
+            t.flush()
+            return backend._get_path(t.name, **kwargs)
 
-# A placeholder for a kwarg default value.
-class NoDefault(object):
-    pass
+    else:
+        BackendError('Backend must provide `_get_path()` or `_get_file()`')
```

Comment:
Shouldn't this also need a `raise`?

Reply:
Modified importing of backends to prevent this outright. Added a warning at import time if a backend does not implement required interface.